src/advanced_unsafety/uninitialized.md

```diff
@@ -0,0 +1,128 @@
+# Uninitialized memory
+
+> _"I'm Nobody! Who are you? Are you — Nobody — too?"_
+>
+> — _Emily Dickinson_
+
+While we have covered [invalid values], there's another thing that behaves a lot like invalid values, but has nothing to do with actual bit patterns: Uninitialized memory.
+
+An easy way to think about uninitialized memory is that there's an additional value (often called `undef` using LLVM's term for it) that does not map to any concrete bit pattern, but can be introduced in abstract in various ways, and makes _most_ values invalid.
+
+If you explicitly wish to work with uninitialized and partially-initialized types, [`MaybeUninit<T>`] is a useful abstraction since it can be "initialized" with no overhead and then written to in parts.
+
+## Sources of uninitialized memory
+
+### `mem::uninitialized()` and `MaybeUninit::assume_init()`
+
+[`mem::uninitialized()`] is a deprecated API that has a very tempting shape, it lets you do things like `let x = mem::uninitialized()` for cases when you want to construct the value in bits. It's basically _always_ UB to use, since it immediately sets `x` to uninitialized memory, which is UB.
+
+Use [`MaybeUninit<T>`] instead.
+
+It is still possible to create uninitialized memory using [`MaybeUninit::assume_init()`] if you have not, in fact, assured that things are initialized.
+
+### Padding
+
+Padding bytes in structs and enums are often but not always uninitialized. This means that treating a struct as a bag of bytes (by, say, treating `&Struct` as `&[u8; size_of::<Struct>()]` and reading from there) is UB even if you don't write invalid values to those bytes, since you are ginning up uninitialized `u8`s.
+
+Reading from padding [always produces uninitialized values][pad-glossary].
+
+
+
+### Moved-from values
+
+The following code is UB:
+
+```rust
+# use std::ptr;
+let x = String::new(); // String is not Copy
+let mut v = vec![];
+let ptr = &x as *const String;
+
+v.push(x); // move x into the vector
+
+unsafe {
+    // reads from moved-from memory
+    let ghost = ptr::read(ptr);
```

Comment:
Isn't the issue here that `ptr` is dangling after `x` is moved, not that the value at `ptr` is necessarily uninitialized?

Reply:
I feel like it's both, right? It can be analyzed through the framework of dangling pointers or through the framework of uninit.

Reply:
I think it is misleading to talk about deallocated and uninitialized memory in the same way. Reads from deallocated memory are insta-UB, whereas reads from uninitialized memory might not be UB if you're reading it as a type such as `MaybeUninit<u8>`.

Reply:
Yeah I'm going to rewrite this section. I'll keep it in since it's _similar_ to uninitialized memory, so worth thinking about, but not the same.

Reply:
I've reworked this a bit, thoughts?

Reply:
I would reword this to start by talking about freed allocations. Then end the section by saying that moving out of a local variable is like deallocating the memory because the compiler might reuse that location on the stack for another local.

Reply:
I'm hesitant to do _too much_ here because I suspect that coverage may live better in a separate chapter. I mentioned freed memory a bit, though.